src/cm/networkmanager/networkmanager.cpp

```diff
@@ -0,0 +1,601 @@
+                    rule.mDstIP   = instance.mIP;
+                    rule.mSrcIP   = ip;
+                    rule.mProto   = protocol.c_str();
+                    rule.mDstPort = port.c_str();
+
+                    return rule;
+                }
+            }
+        }
+    }
+
+    throw std::runtime_error("rule not found");
+}
+
+Error NetworkManager::PrepareFirewallRules(const std::string& subnet, const String& ip,
+    const Array<StaticString<cConnectionNameLen>>& allowedConnections, NetworkParameters& result)
+{
+    if (allowedConnections.IsEmpty()) {
+        return ErrorEnum::eNone;
+    }
+
+    try {
+        for (const auto& connection : allowedConnections) {
+            std::string serviceID, port, protocol;
+            ParseAllowConnection(connection, serviceID, port, protocol);
+            auto rule = GetInstanceRule(serviceID, port, protocol, subnet, ip);
+            result.mFirewallRules.PushBack(rule);
+        }
+    } catch (const std::exception& e) {
+        return Error(aos::ErrorEnum::eRuntime, e.what());
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error NetworkManager::AddHosts(const std::vector<std::string>& hosts, const std::string& ip)
+{
+    for (const auto& host : hosts) {
+        if (IsHostExist(host)) {
+            return Error(ErrorEnum::eAlreadyExist, "host already exists");
+        }
+
+        mHosts[ip].push_back(host);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+void NetworkManager::PrepareInstanceIdentHosts(
+    const InstanceIdent& instanceIdent, const String& networkID, std::vector<std::string>& hosts) const
+{
+    if (instanceIdent.mServiceID.IsEmpty() || instanceIdent.mSubjectID.IsEmpty()) {
+        return;
+    }
+
+    StaticString<cHostNameLen> hostName;
+
+    hostName.Format(
+        "%d.%s.%s", instanceIdent.mInstance, instanceIdent.mSubjectID.CStr(), instanceIdent.mServiceID.CStr());
+    hosts.push_back(hostName.CStr());
+
+    hostName.Format("%d.%s.%s.%s", instanceIdent.mInstance, instanceIdent.mSubjectID.CStr(),
+        instanceIdent.mServiceID.CStr(), networkID.CStr());
+    hosts.push_back(hostName.CStr());
+
+    if (instanceIdent.mInstance == 0) {
+        hostName.Format("%s.%s", instanceIdent.mSubjectID.CStr(), instanceIdent.mServiceID.CStr());
+        hosts.push_back(hostName.CStr());
+    }
+
+    hostName.Format("%s.%s.%s", instanceIdent.mSubjectID.CStr(), instanceIdent.mServiceID.CStr(), networkID.CStr());
+    hosts.push_back(hostName.CStr());
+}
+
+bool NetworkManager::IsHostExist(const std::string& hostName) const
+{
+    for (const auto& [_, hosts] : mHosts) {
+        if (std::find(hosts.begin(), hosts.end(), hostName) != hosts.end()) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void NetworkManager::AddProviderNetwork(
+    const String& networkID, const String& nodeID, NetworkParameters& networkParameters)
+{
+    LOG_DBG() << "Adding provider network" << Log::Field("networkID", networkID) << Log::Field("nodeID", nodeID);
```

Comment:
I expect to see `Added provider network` log. Otherwise, use present simple: `Add provider network`

Reply:
Done